Add sensor platform to LG ThinQ integration

LG-ThinQ-Integration opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add sensor platform to LG ThinQ integration

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/34565

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+AIR_QUALITY_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.PM1: SensorEntityDescription(
+        key=ThinQProperty.PM1,
+        device_class=SensorDeviceClass.PM1,
+        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.PM1,
+    ),
+    ThinQProperty.PM2: SensorEntityDescription(
+        key=ThinQProperty.PM2,
+        device_class=SensorDeviceClass.PM25,
+        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.PM2,
+    ),
+    ThinQProperty.PM10: SensorEntityDescription(
+        key=ThinQProperty.PM10,
+        device_class=SensorDeviceClass.PM10,
+        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.PM10,
+    ),
+    ThinQProperty.HUMIDITY: SensorEntityDescription(
+        key=ThinQProperty.HUMIDITY,
+        device_class=SensorDeviceClass.HUMIDITY,
+        native_unit_of_measurement=PERCENTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.HUMIDITY,
+    ),
```
All these entities with a device class which is not enum (where the name is the same as the device class), can be removed from strings.json and icons.json, as they get their name from the device class

LG-ThinQ-Integration replied:
Updated

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.ODOR,
+        translation_key=ThinQProperty.ODOR,
+    ),
+    ThinQProperty.TOTAL_POLLUTION: SensorEntityDescription(
+        key=ThinQProperty.TOTAL_POLLUTION,
+        translation_key=ThinQProperty.TOTAL_POLLUTION,
+    ),
```
What does this do?

YunseonPark-LGE replied:
TOTAL_POLLUTION has deleted. I'll prepare Enum type and add next sensor PR. Thanks.

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+FILTER_INFO_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.FILTER_LIFETIME: SensorEntityDescription(
+        key=ThinQProperty.FILTER_LIFETIME,
+        native_unit_of_measurement=UnitOfTime.HOURS,
+        translation_key=ThinQProperty.FILTER_LIFETIME,
+    ),
```
device class?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.CURRENT_JOB_MODE,
+        translation_key=ThinQProperty.CURRENT_JOB_MODE,
+    ),
+    ThinQPropertyEx.CURRENT_JOB_MODE_STICK_CLEANER: SensorEntityDescription(
+        key=ThinQProperty.CURRENT_JOB_MODE,
+        translation_key=ThinQPropertyEx.CURRENT_JOB_MODE_STICK_CLEANER,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
It is a job mode for Stick cleaner.


R | OFF | OFF
-- | -- | --
R | NORMAL | NORMAL
R | HIGH | HIGH
R | TURBO | TURBO
R | MOP | MOP
R | AUTO | AUTO

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+LIGHT_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.BRIGHTNESS: SensorEntityDescription(
+        key=ThinQProperty.BRIGHTNESS,
+        translation_key=ThinQProperty.BRIGHTNESS,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.HOOD_OPERATION_MODE,
+        translation_key=ThinQProperty.OPERATION_MODE,
+    ),
+    ThinQProperty.WATER_HEATER_OPERATION_MODE: SensorEntityDescription(
+        key=ThinQProperty.WATER_HEATER_OPERATION_MODE,
+        translation_key=ThinQProperty.OPERATION_MODE,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
This is like HOOD_OPERATION_MODE above.


R | POWER_ON | on state
-- | -- | --
R | POWER_OFF | off state

YunseonPark-LGE replied:
WATER_HEATER_OPERATION_MODE has deleted. I'll prepare next binary_sensor PR. Thanks.

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    )
+}
+PREFERENCE_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.RINSE_LEVEL: SensorEntityDescription(
+        key=ThinQProperty.RINSE_LEVEL,
+        translation_key=ThinQProperty.RINSE_LEVEL,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+RECIPE_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.RECIPE_NAME: SensorEntityDescription(
+        key=ThinQProperty.RECIPE_NAME,
+        translation_key=ThinQProperty.RECIPE_NAME,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.RECIPE_NAME,
+        translation_key=ThinQProperty.RECIPE_NAME,
+    ),
+    ThinQProperty.WORT_INFO: SensorEntityDescription(
+        key=ThinQProperty.WORT_INFO,
+        translation_key=ThinQProperty.WORT_INFO,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.WORT_INFO,
+        translation_key=ThinQProperty.WORT_INFO,
+    ),
+    ThinQProperty.YEAST_INFO: SensorEntityDescription(
+        key=ThinQProperty.YEAST_INFO,
+        translation_key=ThinQProperty.YEAST_INFO,
+    ),
+    ThinQProperty.HOP_OIL_INFO: SensorEntityDescription(
+        key=ThinQProperty.HOP_OIL_INFO,
+        translation_key=ThinQProperty.HOP_OIL_INFO,
+    ),
+    ThinQProperty.FLAVOR_INFO: SensorEntityDescription(
+        key=ThinQProperty.FLAVOR_INFO,
+        translation_key=ThinQProperty.FLAVOR_INFO,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+REFRIGERATION_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.FRESH_AIR_FILTER: SensorEntityDescription(
+        key=ThinQProperty.FRESH_AIR_FILTER,
+        translation_key=ThinQProperty.FRESH_AIR_FILTER,
+    ),
+    ThinQProperty.ONE_TOUCH_FILTER: SensorEntityDescription(
+        key=ThinQProperty.ONE_TOUCH_FILTER,
+        translation_key=ThinQProperty.ONE_TOUCH_FILTER,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
In case of Kimchi refrigerator>

Air filter>
Off / Auto / Power
On touch filter>
On/Off

Btw, I love Kimchi.

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+RUN_STATE_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.CURRENT_STATE: SensorEntityDescription(
+        key=ThinQProperty.CURRENT_STATE,
+        translation_key=ThinQProperty.CURRENT_STATE,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
Current state for the device.
The enum value depends on the device.

In case of Home Brew>

R | STANDBY | Standby
-- | -- | --
R | PREPAREING_FERMENTATION | Preparing
R | DURING_FERMENTATION | Fermenting
R | CARBONATION | Carbonating
R | DURING_AGING | Ripening
R | EXTRACTION_MODE | Storing
R | MELTING | Wort Dissolving
R | TEMPERATURE_STABILIZATION | Temperature Stabilizing
R | EXTRACTING_CAPSULE | Capsule Extracting
R | AS_POP_UP | Error

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.CURRENT_STATE,
+        translation_key=ThinQProperty.CURRENT_STATE,
+    ),
+    ThinQProperty.COCK_STATE: SensorEntityDescription(
+        key=ThinQProperty.COCK_STATE,
+        translation_key=ThinQProperty.COCK_STATE,
+    ),
+    ThinQProperty.STERILIZING_STATE: SensorEntityDescription(
+        key=ThinQProperty.STERILIZING_STATE,
+        translation_key=ThinQProperty.STERILIZING_STATE,
+    ),
```
🤨 
What does this do?

LG-ThinQ-Integration replied:
In Water purifier>

R | CLEANING | Sterilization mode
-- | -- | --
NORMAL | Normal mode

joostlek replied:
And the cock state?

LG-ThinQ-Integration replied:
It's been updated internally.
cock(body) : Cleaning / Normal
sterilization : On/Off/Cancel

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.STERILIZING_STATE,
+        translation_key=ThinQProperty.STERILIZING_STATE,
+    ),
+    ThinQProperty.GROWTH_MODE: SensorEntityDescription(
+        key=ThinQProperty.GROWTH_MODE,
+        translation_key=ThinQProperty.GROWTH_MODE,
+    ),
+    ThinQProperty.WIND_VOLUME: SensorEntityDescription(
+        key=ThinQProperty.WIND_VOLUME,
+        translation_key=ThinQProperty.WIND_VOLUME,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
In Plat cultivator>


R | STANDARD | Standard
-- | -- | --
R | EXT_LEAF | Green Vegetables
R | EXT_HERB | Herbs
R | EXT_FLOWER | Flowers
R | EXT_EXPERT | User Settings

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+TEMPERATURE_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.TARGET_TEMPERATURE: SensorEntityDescription(
+        key=ThinQProperty.TARGET_TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        translation_key=ThinQProperty.TARGET_TEMPERATURE,
+    ),
+    ThinQProperty.DAY_TARGET_TEMPERATURE: SensorEntityDescription(
+        key=ThinQProperty.DAY_TARGET_TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.DAY_TARGET_TEMPERATURE,
+    ),
+    ThinQProperty.NIGHT_TARGET_TEMPERATURE: SensorEntityDescription(
+        key=ThinQProperty.NIGHT_TARGET_TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.NIGHT_TARGET_TEMPERATURE,
+    ),
+    ThinQProperty.TEMPERATURE_STATE: SensorEntityDescription(
+        key=ThinQProperty.TEMPERATURE_STATE,
+        translation_key=ThinQProperty.TEMPERATURE_STATE,
+    ),
+    ThinQProperty.CURRENT_TEMPERATURE: SensorEntityDescription(
+        key=ThinQProperty.CURRENT_TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.CURRENT_TEMPERATURE,
+    ),
```
deivce_class?

LG-ThinQ-Integration replied:
We use Fahrenheit, Celsius in temperature.
As we know, if we use device_class, we have to use only Celsius. right?

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.NIGHT_TARGET_TEMPERATURE,
+    ),
+    ThinQProperty.TEMPERATURE_STATE: SensorEntityDescription(
+        key=ThinQProperty.TEMPERATURE_STATE,
+        translation_key=ThinQProperty.TEMPERATURE_STATE,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
In Plant cultivator>

R | HIGH | High
-- | -- | --
R | NORMAL | Normal

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+WATER_FILTER_INFO_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.USED_TIME: SensorEntityDescription(
+        key=ThinQProperty.USED_TIME,
+        native_unit_of_measurement=UnitOfTime.MONTHS,
+        translation_key=ThinQProperty.USED_TIME,
+    ),
```
device_class?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+WATER_INFO_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.WATER_TYPE: SensorEntityDescription(
+        key=ThinQProperty.WATER_TYPE,
+        translation_key=ThinQProperty.WATER_TYPE,
+    ),
```
What does this do?

LG-ThinQ-Integration replied:
Water type is as belows

COLD | Cold water
-- | --
HOT | Hot water
NORMAL | Normal water
MINERAL | Mineral water
SODA | Soda water

We have the link in the document.
https://thinq.developer.lge.com/en/cloud/docs/thinq-connect/device-profile/water-purifier/

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        translation_key=ThinQProperty.WATER_TYPE,
+    ),
+}
+TIMER_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
```
When are these used? Looks like a lot of entities

LG-ThinQ-Integration replied:
Unfortunately, we have lots of device variations. Because those are home appliances.
And the devices have lots of several types of timer. ex) timer to start, timer to stop, timer to sleep, timer for remained etc.
Those are split into relative timer, absolute timer.
That's why we have lots of entities.

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        """Initialize a sensor entity."""
+        super().__init__(coordinator, entity_description, property_id)
+
+        self._attr_options = self.data.options
+        if self.options:
+            self._attr_device_class = SensorDeviceClass.ENUM
```
Oh, you already set options. Ideally we have them defined in the entity descriptions, unless they are dynamic

LG-ThinQ-Integration replied:
Because it is dynamic, we can't add it in description.

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+            self._attr_native_unit_of_measurement = None
+            return
+
+        unit_of_measurement = self.data.unit or self.native_unit_of_measurement
+        if self.native_unit_of_measurement != unit_of_measurement:
+            self._attr_native_unit_of_measurement = unit_of_measurement
```
the native unit of measurement ideally doesn't change, can you elabroate?

LG-ThinQ-Integration replied:
If a user change unit_of_measurement, the cloud send us the data with the unit.
For example, a user can choose Fahrenheit or Celsius in the refrigerator.

YunseonPark-LGE replied:
We use Celsius for default native_unit_of_measurement. If the data.unit(dynamic per device) exist and differ to current unit, then the native_unit_of_measurement will be changed.

joostlek reviewed (commented):
Okay I put down a lot of things because we have to find out how every entity is best displayed. so this will be a lot of talking :)

LG-ThinQ-Integration commented:
> Okay I put down a lot of things because we have to find out how every entity is best displayed. so this will be a lot of talking :)

Awesome!!
You are so fast.

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQPropertyEx.ERROR,
+        translation_key=ThinQPropertyEx.ERROR,
+    ),
+    ThinQPropertyEx.NOTIFICATION: SensorEntityDescription(
+        key=ThinQPropertyEx.NOTIFICATION,
+        translation_key=ThinQPropertyEx.NOTIFICATION,
+    ),
```
Are you asking what notification is?
Then there are sensory entity for notification.
ThinQ cloud will give us notification and the entity will display it as string.

joostlek replied:
What kind of notifications do we get? On what kind of devices do we get it?

LG-ThinQ-Integration replied:
example>
Push Code | Description
-- | --
DRYING_IS_COMPLETE | Drying is complete.
DRYING_FAILED | Drying has failed.

LG-ThinQ-Integration replied:
example2>
Value | Description
-- | --
CLEANING_IS_COMPLETE | Cleaning is complete.
ERROR_DURING_CLEANING | Cleaning stopped due to an error.
WATER_LEAK_HAS_OCCURRED | Water leakage has occurred.
RINSE_IS_NOT_ENOUGH | There is not enough rinse.
SALT_REFILL_IS_NEEDED | Salt refill is required.

joostlek replied:
Oh and how many are possible? I will have to open your list probably, but I am expectign a lot (trying to define the scope so we can make a good decision)

LG-ThinQ-Integration replied:
Notification is only provided as event.
Removed.

joostlek replied:
That's awesome :D

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+from .entity import ThinQEntity
+
+COMMON_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQPropertyEx.ERROR: SensorEntityDescription(
+        key=ThinQPropertyEx.ERROR,
+        translation_key=ThinQPropertyEx.ERROR,
+    ),
```
There are sensor entities for errors.
ThinQ cloud will give them erros(not communication error but a device's physical error) and the entities will display them.
- https://thinq.developer.lge.com/en/cloud/docs/thinq-connect/device-profile/Dryer/#4

joostlek replied:
So the state of this entity would be `Temperature sensor error`?

LG-ThinQ-Integration replied:
To be specific, a home appliance device could have a sensor entity for various error description.

LG-ThinQ-Integration replied:
To make you understand more, let me give an example.


https://thinq.developer.lge.com/en/cloud/docs/thinq-connect/device-profile/dish-washer/

Dish-Washer

-- | --
WATER_LEAKAGE_ERROR | The error on water leakage
BUBBLE_ERROR | The error on the bubble
HEATER_CIRCUIT_ERROR | The error on the heater circuit
WATER_SUPPLY_ERROR | The error on water supply
MOTOR_ERROR | The error on motor
WATER_DRAIN_ERROR | The error on water drainage
TEMPERATURE_SENSOR_ERROR | The error on the temperature sensor

joostlek replied:
How big would this list be? Ideally we'd have all the (relevant for the appliance) options in the options and have the ability to translate them

YunseonPark-LGE replied:
11 device types have error, each of them is 7 ~ 15 states.
We are not fully prepared the all error states, so just set value to lower case. ex) bubble_error, heater_circuit_error

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.TOTAL_POLLUTION,
+        translation_key=ThinQProperty.TOTAL_POLLUTION,
+    ),
+    ThinQProperty.MONITORING_ENABLED: SensorEntityDescription(
+        key=ThinQProperty.MONITORING_ENABLED,
+        translation_key=ThinQProperty.MONITORING_ENABLED,
+    ),
```
For Air-Quality sensor of Air-Purifier

monitoringEnabled(Sensor Monitoring On) | Enum | R | ON_WORKING | Sensor Monitoring - During Operation
-- | -- | -- | -- | --
R | ALWAYS | Sensor Monitoring - Always

joostlek replied:
So it means if it is monitoring during operation or always?

LG-ThinQ-Integration replied:
Exactly.

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+BATTERY_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.BATTERY_LEVEL: SensorEntityDescription(
+        key=ThinQProperty.BATTERY_LEVEL,
+        translation_key=ThinQProperty.BATTERY_LEVEL,
+    ),
```
R | HIGH | HIGH
-- | -- | --
R | MID | MID
R | LOW | LOW
R | WARNING | WARNING

LG-ThinQ-Integration replied:
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+DETERGENT_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.DETERGENT_SETTING: SensorEntityDescription(
+        key=ThinQProperty.DETERGENT_SETTING,
+        translation_key=ThinQProperty.DETERGENT_SETTING,
+    ),
```
Wash-Tower>

Profile | AUTO | Automatic Setup Devices
-- | -- | --
Profile | NORMAL | Non-Automatic Setup Devices

LG-ThinQ-Integration replied:
Removed

joostlek commented:
I think the way we can proceed with this PR is to see which entities are ready to go, and I suggest we split off the ones we still have to zoom in on, so we at least have the majority merged. This way the whole PR doesn't get blocked by the individual entities. Do you agree on that or do you have another preference?

LG-ThinQ-Integration commented:
After I summarize what you mentioned, Let me give you the plan about the next step(split and etc)
1. Use device_class and enum if possible
=> We can use num in some cases, but in other cases like ODOR, we just get the number. So we can't use enum
2. You want to know about the several property.
=> Let me explain after split.
3. Use event instead of sensor for 'Notificaton'
=> Actually we are planning to add Event entity later. And for some reasons we wanted to support sensor and event entities for notification. But I think it is quite unnecessary. So we decided to remove Notification sensor entities.
4. Split into PRs
=> I agree, there are too many entities. So, I will update the PR which has less than half of the current count.

After I update everything. I will let you know.
And let me change the PR draft.

LG-ThinQ-Integration commented:
I reduced the number of entities(types) from 21 to 9 for this PR.
The entities below are included.
----------------------------------------
ERROR_DESC
AIR_QUALITY_SENSOR_DESC
HUMIDITY_SENSOR_DESC
JOB_MODE_SENSOR_DESC
OPERATION_SENSOR_DESC
REFRIGERATION_SENSOR_DESC
RUN_STATE_SENSOR_DESC
TEMPERATURE_SENSOR_DESC
WATER_INFO_SENSOR_DESC
----------------------------------------

LG-ThinQ-Integration commented:
Let us start from here again.
Let me answer all your current reviews.
If some are not included here, let me comment like 'Not included'
I hope these are not too many for you. :)

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        state_class=SensorStateClass.MEASUREMENT,
+        translation_key=ThinQProperty.HUMIDITY,
+    ),
+    ThinQProperty.ODOR: SensorEntityDescription(
+        key=ThinQProperty.ODOR,
+        translation_key=ThinQProperty.ODOR,
+    ),
```
ODR is not enum but number.
Because we get number from the cloud, we don't have any enum information.

joostlek replied:
Are the numbers 0-4 set in stone? I would recommend transforming that using a dict
```py
{
    0: "weak"
}
```
And then we can provide a list of options to the entity description

LG-ThinQ-Integration replied:
We don't use that way in other environments.
And I am not sure it has small range of numbers.

joostlek replied:
This could be a dictionary in the HA code, but this way we transform it to something user readable, because the untranslated values are still used in templating, so I want to avoid using numbers

YunseonPark-LGE replied:
Odor property has not yet supported enum llist. I'll remove it's state translation in strings.json.
        "state": {
          "1": "Weak",
          "2": "Normal",
          "3": "Strong",
          "4": "Very strong"
        }

joostlek replied:
We do want these translations, but we should transform them to snake case strings

YunseonPark-LGE replied:
I have question. Does this mean can't assign any numbers to keys in strings.json? ex) "1234"

joostlek replied:
Ideally we don't do this as that makes the state unclear for people who are using templates to automate. Having a proper name also helps people who translate

YunseonPark-LGE replied:
ODOR has deleted. I'll prepare Enum type and add next sensor PR. Thanks.

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.BATTERY_PERCENT,
+        native_unit_of_measurement=PERCENTAGE,
+        translation_key=ThinQProperty.BATTERY_PERCENT,
+    ),
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ThinQProperty.DOOR_STATE: SensorEntityDescription(
+        key=ThinQProperty.DOOR_STATE,
+        translation_key=ThinQProperty.DOOR_STATE,
+    ),
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ),
+}
+OPERATION_SENSOR_DESC: dict[ThinQProperty, SensorEntityDescription] = {
+    ThinQProperty.HOOD_OPERATION_MODE: SensorEntityDescription(
+        key=ThinQProperty.HOOD_OPERATION_MODE,
+        translation_key=ThinQProperty.OPERATION_MODE,
+    ),
```
I know what you are thinking.
You might think that we'd better handle it as binary sensor.
But because the cloud sent it as string based on the profile spec as below.
- https://thinq.developer.lge.com/en/cloud/docs/thinq-connect/device-profile/Hood/
We need to handle it as sensor.



R | POWER_ON | Hood On
-- | -- | --
R | POWER_OFF | Hood Off

joostlek replied:
We will have more binary sensors right? How do they receive their state? I think we should explore the possibility to add them as binary sensor. I'm quite convinced that that is possible. So let's also split this one off. If we find out it's indeed going to be a big hassle, we can always decide to merge it in again

YunseonPark-LGE replied:
HOOD_OPERATION_MODE has deleted. I'll prepare next binary_sensor PR. Thanks.

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ThinQProperty.POWER_LEVEL: SensorEntityDescription(
+        key=ThinQProperty.POWER_LEVEL,
+        translation_key=ThinQProperty.POWER_LEVEL,
+    )
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ThinQProperty.CLEAN_LIGHT_REMINDER: SensorEntityDescription(
+        key=ThinQProperty.CLEAN_LIGHT_REMINDER,
+        translation_key=ThinQProperty.CLEAN_LIGHT_REMINDER,
+    ),
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ThinQProperty.SIGNAL_LEVEL: SensorEntityDescription(
+        key=ThinQProperty.SIGNAL_LEVEL,
+        translation_key=ThinQProperty.SIGNAL_LEVEL,
+    ),
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ThinQProperty.MACHINE_CLEAN_REMINDER: SensorEntityDescription(
+        key=ThinQProperty.MACHINE_CLEAN_REMINDER,
+        translation_key=ThinQProperty.MACHINE_CLEAN_REMINDER,
+    ),
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ThinQProperty.SOFTENING_LEVEL: SensorEntityDescription(
+        key=ThinQProperty.SOFTENING_LEVEL,
+        translation_key=ThinQProperty.SOFTENING_LEVEL,
+    ),
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.DURATION,
+        native_unit_of_measurement=UnitOfTime.HOURS,
+        translation_key=ThinQProperty.DURATION,
+    ),
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+    ThinQProperty.CURRENT_DISH_WASHING_COURSE: SensorEntityDescription(
+        key=ThinQProperty.CURRENT_DISH_WASHING_COURSE,
+        translation_key=ThinQProperty.CURRENT_DISH_WASHING_COURSE,
+    )
```
Removed

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,590 @@
+        key=ThinQProperty.CURRENT_JOB_MODE,
+        translation_key=ThinQPropertyEx.CURRENT_JOB_MODE_STICK_CLEANER,
+    ),
+    ThinQProperty.PERSONALIZATION_MODE: SensorEntityDescription(
+        key=ThinQProperty.PERSONALIZATION_MODE,
+        translation_key=ThinQProperty.PERSONALIZATION_MODE,
+    ),
```
It is like preset mode.

In case of Air-purifier

baby, cook mode.

joostlek commented:
I already responded to some of them, I will continue when I'm at my desk again.

I saw that you were asking about device class and the native unit of measurement. Ideally everything that can have a device class, has one. The benefit of having a device class is that it allows the user to update the unit of measurement in the frontend. So ideally the unit we get is always the same, and HA does the conversion for you

LG-ThinQ-Integration commented:
> I already responded to some of them, I will continue when I'm at my desk again.
> 
> I saw that you were asking about device class and the native unit of measurement. Ideally everything that can have a device class, has one. The benefit of having a device class is that it allows the user to update the unit of measurement in the frontend. So ideally the unit we get is always the same, and HA does the conversion for you

About the native unit of measurement,
If a user control it in the HA UI, but if the user control it outside HA, like ThinQ smartpone application,
The cloud send it with the different unit.
So we need to handle like this.

joostlek commented:
so the API responds with what the user set as unit of measurement?

joostlek commented:
Also, since this will be a lot of discussion, I think we can start doing stuff simultaneous and get some other platforms in :)

LG-ThinQ-Integration commented:
> so the API responds with what the user set as unit of measurement?

Exactly~!!

joostlek commented:
For what datapoints is this the case?

LG-ThinQ-Integration commented:
> For what datapoints is this the case?

What do you mean?

joostlek commented:
What entities have this changing unit of measurement?

LG-ThinQ-Integration commented:
> What entities have this changing unit of measurement?

Temperature related entities.
Current temperature, target temperature ...

LG-ThinQ-Integration commented:
> Also, since this will be a lot of discussion, I think we can start doing stuff simultaneous and get some other platforms in :)
Do you mean  I can change  the PR open?

joostlek commented:
You can open a second PR

joostlek commented:
Check, that means we need to have logic to return the right native UoM for those entities

LG-ThinQ-Integration commented:
> You can open a second PR
 
As I understand,
this PR will be merged.
And I need to make a second sensor PR.
Right?

joostlek commented:
Oh no I mean, you can maybe already start looking at a second platform PR like binary_sensor or cover or something like that. We can only add more sensors after this one has been merged

LG-ThinQ-Integration commented:
> Oh no I mean, you can maybe already start looking at a second platform PR like binary_sensor or cover or something like that. We can only add more sensors after this one has been merged

Binary sensor is already merged.
Please tell me what I have to do more?

joostlek commented:
Oh lol, forgot that we already did that. I mean, you had a massive PR at the beginning, at this point we can open multiple PRs at the same time to keep pace :)

joostlek commented:
Okay so you now removed the entities with translatable state. So the things left for this PR to be ready is deciding what we do with the binary sensors and the native value for temperature entities?

joostlek commented:
https://github.com/joostlek/core/pull/new/lg-thinq-binary I think we can use something like this to make some of them a binary sensor

YunseonPark-LGE commented:
Thanks for your suggestion, binary sensor will fix by modifying data.is_on's condition.
self._attr_is_on = self.data.is_on

And please point out what is "the native value for temperature entities?"

joostlek commented:
Oh sorry, I meant native unit of measurement. I understood that the native unit of measurement can change, and the entity should act on that accordingly, otherwise it will display wrong values

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,315 @@
+        )
+        self._attr_native_value = self.data.value
+
+        # Update unit.
+        if self.device_class == SensorDeviceClass.ENUM:
+            self._attr_native_unit_of_measurement = None
+            return
```
why do we have this check?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,315 @@
+            self._attr_native_unit_of_measurement = None
+            return
+
+        unit_of_measurement = self.data.unit or self.native_unit_of_measurement
```
What does the `data.unit` return exactly? When is this set?

LG-ThinQ-Integration replied:
In temperature, it could be one of Fahreheit or Celsius.

joostlek reviewed (changes requested):
Oh I didn't see that the function was already added to update the native UoM, sorry

joostlek commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,315 @@
+        native_unit_of_measurement=PERCENTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    ThinQProperty.MONITORING_ENABLED: SensorEntityDescription(
+        key=ThinQProperty.MONITORING_ENABLED,
+        translation_key=ThinQProperty.MONITORING_ENABLED,
+    ),
```
Would it make sense to extend `SensorEntityDescription` to have a value_fn?

```suggestion
    ThinQProperty.MONITORING_ENABLED: SensorEntityDescription(
        key=ThinQProperty.MONITORING_ENABLED,
        translation_key=ThinQProperty.MONITORING_ENABLED,
        value_fn=lambda value: value.lower(),
        options=["on_working", "always"],
    ),
```

This would transform it to `on_working` and `always`. For the number enums you can do
```py
ODOR_MAPPING = {
    0: "bad",
}
```

```py
    ThinQProperty.MONITORING_ENABLED: SensorEntityDescription(
        key=ThinQProperty.MONITORING_ENABLED,
        translation_key=ThinQProperty.MONITORING_ENABLED,
        value_fn=lambda value: ODOR_MAPPING.get(value),
        options=list(ODOR_MAPPING.values()),
    ),
```

How does this sound?

LG-ThinQ-Integration replied:
We used the translation with string.json.

LG-ThinQ-Integration commented:
> Oh sorry, I meant native unit of measurement. I understood that the native unit of measurement can change, and the entity should act on that accordingly, otherwise it will display wrong values

Exactly.

joostlek commented:
By the way, compliments to you and your team on the product, I've just seen the stand and it looks amazing ❤️

![IMG_20240909_114935823.jpg](https://github.com/user-attachments/assets/ecef9bd2-4cf3-4890-9085-adf0aa43fff1)

LG-ThinQ-Integration commented on homeassistant/components/lg_thinq/sensor.py:
```diff
@@ -0,0 +1,338 @@
+            and self.data.value not in self.options
+        ):
+            # For data is not in options, unusual temperature's numeric data
+            self.entity_description.options_ex_fn(self._attr_options, self.data.value)
```
v

LG-ThinQ-Integration replied:
Removed

LG-ThinQ-Integration commented:
I guess I commented every review you made.
If anything is missing, please let me know.
We added 'state' in 'string.json' for translation and updated login for UoM

This needs to be rebased after the bump to 0.9.7 PR is merged.

LG-ThinQ-Integration commented:
> By the way, compliments to you and your team on the product, I've just seen the stand and it looks amazing ❤️
> 
> ![IMG_20240909_114935823.jpg](https://private-user-images.githubusercontent.com/7083755/365618720-ecef9bd2-4cf3-4890-9085-adf0aa43fff1.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU4ODYwMDAsIm5iZiI6MTcyNTg4NTcwMCwicGF0aCI6Ii83MDgzNzU1LzM2NTYxODcyMC1lY2VmOWJkMi00Y2YzLTQ4OTAtOTA4NS1hZGYwYWE0M2ZmZjEuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDlUMTI0MTQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDdiOWMwNjFkYWFjYWZkNTVjZDFlMGE2NTYyMzcyZDdhZDZiYTM4YTRjZjIwZGFjN2ZlMmVlZmEyNzI0OGFiNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7n-sjNVIKCRhL7E_sUsHnTTK_b8gRNdoq7iVz1lHI9Y)

Thank you very much~!! :)

LG-ThinQ-Integration commented:
Btw, After this is merged, I am planning as below.
1. mqtt PR
2. 3 PRs together
- add remained sensor entities PR
- add climate platform PR
- add binary sensor entities moving from sensor PR

LG-ThinQ-Integration commented:
I rebased this PR.

LG-ThinQ-Integration commented:
I guess I left comments for all your reviews.
If not, please let me know.
Btw, is this PR waiting for other PRs? Or after you review, is it going to be merged?

joostlek commented on homeassistant/components/lg_thinq/strings.json:
```diff
@@ -43,6 +43,371 @@
+          "temperature_number": "Number"
+        }
+      },
+      "one_touch_filter": {
+        "name": "Fresh air filter",
+        "state": {
+          "off": "Off",
+          "on": "On"
+        }
+      }
```
Why wasnt this a binary sensor?

LG-ThinQ-Integration replied:
Removed

joostlek commented:
I am asking some second opinions to make sure we solved all the blockers in the PR. That's why I already reviewed the rest for a bit :)

LG-ThinQ-Integration commented:
Rebased.
And let me check your reviews.

joostlek commented:
No I don't have more comments, but I want a second pair of eyes to make sure this is ready to merge

JangwonLeeLG committed:
Add sensor platform to LG ThinQ integration

LG-ThinQ-Integration commented:
Because mqtt is merged, I changed a little bit on coordinator access in runtime_data.

joostlek committed:
Dedup

joostlek committed:
Dedup

joostlek committed:
Dedup

joostlek committed:
Dedup

joostlek committed:
Dedup

joostlek committed:
Fix water type enum

joostlek committed:
Dedup
